extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
+                    //    details, which is effectively schema1.toString().equals(schema2.toString()).
+                    // 2. There must be a 1-to-1 mapping between Schema's and their string representations
+                    //    for bidirectional lookup.
+                    String schemaJson = schemaToJson.computeIfAbsent(datum.getSchema(), Schema::toString);
+                    out.writeString(schemaJson);
+                    out.writeByteArray(datum instanceof LazyImmutableContainer
+                            ? ((LazyImmutableContainer<?>) datum).serialized
+                            : serialize(new GenericDatumWriter<>(datum.getSchema()), datum));
+                }
+
+                @Nonnull
+                @Override
+                public GenericContainer read(@Nonnull ObjectDataInput in) throws IOException {
+                    String schemaJson = in.readString();
+                    Schema schema = jsonToSchema.computeIfAbsent(schemaJson,
+                            json -> new Schema.Parser().parse(json));
+                    byte[] serializedDatum = in.readByteArray();
+                    switch (schema.getType()) {
+                        case RECORD:
+                            return new LazyImmutableRecord(serializedDatum, schema);
+                        case ARRAY:
+                            return new LazyImmutableArray<>(serializedDatum, schema);
+                        case ENUM:  // GenericEnumSymbol
+                        case FIXED: // GenericFixed
+                            return deserialize(new GenericDatumReader<>(schema), serializedDatum);
+                        default:
+                            throw new UnsupportedOperationException("Schema type " + schema.getType() + " is unsupported");
+                    }
+                }
+            };
+        }
 
         @Override
-        public Class<Utf8> getSerializationType() {
-            return Utf8.class;
+        public boolean isOverwritable() {
+            return false;
+        }
+    }
+
+    /**
+     * Replaces {@link Utf8} with {@link String} globally. Alternatively,
+     * the string class can be set on a per-schema basis via <ol>
+     * <li> {@link GenericData#setStringType GenericData.setStringType}{@code (schema,}
+     *      {@link StringType#String String}{@code )} or,
+     * <li> {@code SchemaBuilder.record("name").prop(}{@link GenericData#STRING_PROP STRING_PROP}{@code ,}
+     *      {@link StringType#String String}{@code )}.
+     */
+    public static final class CharSequenceHook implements SerializerHook<CharSequence> {
+        @Override
+        public Class<CharSequence> getSerializationType() {
+            return CharSequence.class;
         }
 
         @Override
         public Serializer createSerializer() {
-            return new StreamSerializer<Utf8>() {
+            return new ByteArraySerializer<CharSequence>() {
```

Comment:
can this have any impact on backwards compatibility?

these serializers might not only be used by Jet, so what would happen if someone relied on them, eg. kept avro-serialized data in IMap?

Reply:
Writing only a byte array in a `StreamSerializer` is equivalent to returning that array in a `ByteArraySerializer`. You can look at `StreamSerializerAdapter` and `ByteArraySerializerAdapter`.

Reply:
I can see that you reversed this change to use Utf8 again